Zend/zend_execute.c

```diff
@@ -2354,6 +2354,9 @@ static zend_always_inline void zend_fetch_dimension_address(zval *result, zval *
 				ZVAL_UNDEFINED_OP1();
 			}
 			if (type != BP_VAR_UNSET) {
+				if (Z_TYPE_P(container) == IS_FALSE) {
+					zend_error(E_DEPRECATED, "Automatic conversion of false to array is deprecated");
+				}
```

Comment:
Shouldn't the BP_VAR_UNSET case also throw a deprecation, though possibly not with the same message? `unset(true[0])` is an error condition, so `unset(false[0])` should also become one.

Reply:
I have not been able to identify a test case when `unset()` could be used with autovivification. The example you gave throws an error
>Cannot use temporary expression in write context

If I use a variable then there is no error. 

```
$a = true;
unset($a[0]);
```
The unset is a no-op. 

And of course, there is a different error for this
```
$a = false;
unset($a[]);
```
> Cannot use [] for unsetting

So when would `unset()` use autovivification?

Reply:
I have no idea how JIT works or what I need to change there. I have added the deprecation message to JIT test case, but it didn't require any changes in the C code.

Reply:
> I have not been able to identify a test case when `unset()` could be used with autovivification. The example you gave throws an error
> 
> > Cannot use temporary expression in write context
> 
> If I use a variable then there is no error.
> 
> ```
> $a = true;
> unset($a[0]);
> ```
> 
> The unset is a no-op.

Huh, I expected that to be an error. This variant does error:
```
<?php 
$a = true;
unset($a[0][0]);
```

The fact that it doesn't error without nesting is probably an implementation bug -- I'll have to look a bit closer.

Reply:
I've fixed the missing error in https://github.com/php/php-src/commit/b64213872a15136a088a50035673b2a7ee90a6d2.